x-pack/plugins/security_solution/common/api/detection_engine/prebuilt_rules/model/diff/three_way_diff/three_way_diff.ts

```diff
@@ -98,6 +104,13 @@ export interface ThreeWayDiff<TValue> extends ThreeVersionsOf<TValue> {
    */
   merge_outcome: ThreeWayMergeOutcome;
 
+  /**
+   * Boolean which determines if a base version was found and returned for the three-way-diff of the field
+   * - true: the base version of the field was found and is either defined or undefined
+   * - false: the base version of the field was not found
+   */
+  has_base_version: boolean;
```

Comment:
GitHub doesn't allow to post a comment on line 36, so posting it here.

```ts
export interface ThreeVersionsOf<TValue> {
  base_version: TValue | MissingVersion;
  ...
```

What are your thoughts on removing the `MissingVersion` type from `ThreeVersionsOf` and just relying on newly added `has_base_version` boolean?

Pros of removing:
 - A bit type safer because FE uses the same `ThreeWayDiff` type. And `ThreeWayDiff` says that `base_version` is `TValue | MissingVersion`, but on FE it'll actually be `TValue | undefined`.
 - One way to check if base version is missing (`has_base_version`) on backend vs two ways (`has_base_version` and `baseVersion === MissingValue`).

Pros of leaving as is:
 - We won't need to update the code in a bunch of places :)

Reply:
Not sure I get your proposal: what's your idea for the type? `base_version: TValue`?

In the backend we still need to distinguish the case where the base_version field exists and is undefined vs the case where the base_version does not exist. And we need to do certain checks such as:

https://github.com/elastic/kibana/blob/9faf2f6ebf06941445a05e2a9b63ea374cbcb7f8/x-pack/plugins/security_solution/server/lib/detection_engine/prebuilt_rules/logic/diff/calculation/calculate_rule_fields_diff.ts#L61

I.e: if we remove the possibility of `base_version` being MissingVersion, how we differentiate both cases?

Reply:
I see why this is confusing. We use `MissingValue` to indicate one of two things:
 - base_version of a rule is missing
 - field value is missing because the base_version of a rule is missing (my idea was to remove `MissingValue` for this use case)

In the example you referred to (`calculate_rule_fields_diff.ts:61`) `MissingVersion` means "base_version of a rule is missing". I think it's okay and should stay as is.

But we could probably kick the `MissingVersion` out of `ThreeWayDiff`, where it means "field value is missing because the base_version of a rule is missing". We'll just have
```ts
interface ThreeWayDiff<TValue> {
 ...
 has_base_version: boolean;
 base_version?: TValue;
 ...
}
```

Reply:
I just understood that then we can't change it in `ThreeVersionsOf` because it's used for both use cases. Then we can just put these fields directly ThreeWayDiff type definition without using `extends ThreeVersionsOf`.

Smth like:
```ts
interface ThreeWayDiff<TValue> {
  base_version?: TValue;
  current_version: TValue;
  target_version: TValue;
  // ... all the other ThreeWayDiff fields
}
``

Hope it makes sense

Reply:
@jpdjere `MissingVersion` is a `Symbol` and it is not serialized into JSON when returned from the API. This means that we should stop relying on it and should remove from the API contract. Both `ThreeWayDiff<TValue>` and `ThreeVersionsOf<TValue>` types is a part of the API contract, so we should refactor them in a way so they don't depend on `MissingVersion` anymore. One way to do it would be:

Reply:
And we probably should remove the `MissingVersion` constant from the codebase completely, to not have two ways of defining a missing base version.

Reply:
Another way to do it would be to remove inheritance from `ThreeWayDiff<TValue> extends ThreeVersionsOf<TValue>`:

```ts
export interface ThreeVersionsOf<TValue> {
  base_version: TValue | MissingVersion;
  current_version: TValue;
  target_version: TValue;
}
```

```ts
export interface ThreeWayDiff<TValue> {
  has_base_version: boolean;
  base_version: TValue | undefined;
  current_version: TValue;
  target_version: TValue;
  merged_version: TValue;
  diff_outcome: ThreeWayDiffOutcome;
  merge_outcome: ThreeWayMergeOutcome;
  has_update: boolean;
  conflict: ThreeWayDiffConflict;
}
```

If we do it, `ThreeVersionsOf<TValue>` will be removed from the API contract which means we could continue using `MissingVersion` in it.

Reply:
I think, from a diff writing perspective, it could be easier without the `MissingVersion` symbol in there just for typing sake but overall the logic would remain the same. Given the way we calculate `MissingVersion` [here](https://github.com/elastic/kibana/blob/2c8b2ff4a55756a345b0124f2beaef7a5ce849c9/x-pack/plugins/security_solution/server/lib/detection_engine/prebuilt_rules/logic/diff/calculate_rule_diff.ts#L73), I think it'd be a pretty easy removal if we add that extra `has_base_version` field

Reply:
Removed inheritance of `ThreeWayDiff`. Thanks for the proposals 👍